src/v/bytes/details/io_fragment.h

```diff
@@ -94,14 +102,273 @@ class io_fragment {
+            }
+            return *this;
+        }
+        iter operator++(int) {
+            auto tmp = *this;
+            ++*this;
+            return tmp;
+        }
+        iter& operator--() {
+            if constexpr (Forward) {
+                if (_current == nullptr) {
+                    _current = _container->_tail;
+                } else {
+                    _current = _current->_prev;
+                }
+            } else {
+                if (_current == nullptr) {
+                    _current = _container->_head;
+                } else {
+                    _current = _current->_next;
+                }
+            }
+            return *this;
+        }
+        iter operator--(int) {
+            auto tmp = *this;
+            --*this;
+            return tmp;
+        }
+        bool operator==(const iter& o) const { return _current == o._current; }
+        bool operator!=(const iter& o) const { return _current != o._current; }
+
+    private:
+        friend class io_fragment_list;
+
+        inline void check_generation() const {
+#ifndef NDEBUG
+            vassert(
+              _container->_generation == _my_generation,
+              "Attempting to use an invalidated iterator. The corresponding "
+              "iobuf has been mutated since this iterator was constructed.");
+#endif
+        }
+
+        pointer _current = nullptr;
+#ifndef NDEBUG
+        size_t _my_generation = 0;
+#endif
+        container_type* _container = nullptr;
+    };
+
+public:
+    io_fragment_list() noexcept = default;
+    io_fragment_list(io_fragment_list&& o) noexcept
+      : _head(std::exchange(o._head, nullptr))
+      , _tail(std::exchange(o._tail, nullptr)) {}
+    io_fragment_list(const io_fragment_list&) = delete;
+    io_fragment_list& operator=(io_fragment_list&&) = delete;
+    io_fragment_list& operator=(const io_fragment_list&) = delete;
+    ~io_fragment_list() noexcept {
+        vassert(
+          _head == nullptr && _tail == nullptr,
+          "io_fragment_list cleanup failure and possible memory leak");
+    };
+
+    using iterator = iter<false, true>;
+    using reverse_iterator = iter<false, false>;
+    using const_iterator = iter<true, true>;
+
+    bool empty() const {
+        bool e = _head == nullptr;
+        assert(e == (_tail == nullptr));
+        return e;
+    }
+
+    io_fragment& front() {
+        assert(_head != nullptr);
+        return *_head;
+    }
+    io_fragment& front() const {
+        assert(_head != nullptr);
+        return *_head;
+    }
+
+    void pop_front() {
+        assert(_head != nullptr);
+        assert(_head->_prev == nullptr);
+        io_fragment* h = _head;
+        _head = h->_next;
+        h->_next = nullptr;
+        if (_head != nullptr) {
+            _head->_prev = nullptr;
+        } else {
+            _tail = nullptr;
+        }
+        update_generation();
+    }
+
+    void push_front(io_fragment& elem) {
```

Comment:
Are we guaranteed that elem._prev is `nullptr` here? It is required for consistency, maybe we should assert it?